litestar/contrib/pydantic/utils.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+from litestar.constants import UNDEFINED_SENTINELS
+from litestar.types import Empty
+from litestar.utils import is_class_and_subclass
+
+# isort: off
+try:
+    from pydantic import v1 as pydantic_v1
+    import pydantic as pydantic_v2
+except ImportError:
+    try:
+        import pydantic as pydantic_v1  # type: ignore[no-redef]
+
+        pydantic_v2 = Empty  # type: ignore[assignment]
+
+    except ImportError:  # pyright: ignore
+        pydantic_v1 = Empty  # type: ignore[assignment]
+        pydantic_v2 = Empty  # type: ignore[assignment]
+# isort: on
+
+
+if TYPE_CHECKING:
+    from typing_extensions import TypeGuard
+
+PYDANTIC_UNDEFINED_SENTINELS = set()
+
+if pydantic_v1 is not Empty:  # type: ignore[comparison-overlap]
+    PYDANTIC_UNDEFINED_SENTINELS.add(pydantic_v1.fields.Undefined)  # pyright: ignore
+    if pydantic_v2 is not Empty:  # type: ignore[comparison-overlap]
+        PYDANTIC_UNDEFINED_SENTINELS.add(pydantic_v2.fields.PydanticUndefined)  # type: ignore[attr-defined]
+
+UNDEFINED_SENTINELS.update(PYDANTIC_UNDEFINED_SENTINELS)
```

Comment:
Not sure if this is a good idea actually. This only modifies that global when its imported, which is guaranteed to happen if you're using the Pydantic plugin, but it is a side effect.

On the other hand, I don't really like having the Pydantic stuff in the regular namespace, but maybe we should take care of that with a separate PR

Reply:
Yeah I agree. I think it would be really nice if both `PYDANTIC_UNDEFINED_SENTINELS` and `UNDEFINED_SENTINTELS` where instances of `frozenset`.

Reply:
I'm thinking of leaving it like this for now, and then do a larger refactoring that removes the need for this alltogether. Wdyt?

Reply:
@provinzkraut, sorry I missed this. I think leaving it for a later refactoring is a good idea as well to keep this PR focused.